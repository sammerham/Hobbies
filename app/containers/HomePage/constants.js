export const GET_HOBBIES_ASYNC = 'GET_HOBBIES_ASYNC';
export const GET_HOBBIES = 'GET_HOBBIES';
export const GET_REQUEST_FAILED = 'GET_REQUEST_FAILED';
