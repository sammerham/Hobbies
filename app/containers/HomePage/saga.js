import { put, takeEvery, call } from 'redux-saga/effects';
import axios from 'axios';
import {
  GET_HOBBIES_ASYNC,
  GET_HOBBIES,
  GET_REQUEST_FAILED,
} from './constants';

// worker saga: makes the api call when watcher saga sees the GET_HOBBIES action
function* getHobbiessAsync() {
  /**  
   * fetch hobbies from api
   * @params
   * None
   * @returns 
  {
    "hobbies":
    [
      "running",
      "reading",
      "walking",
      "gardening"
    ]
  }
   */
  try {
    // const hobbies = yield getHHobbies();
    const result = yield call(axios.get, '/api');
    const { hobbies } = result.data;

    // dispatches GET_HOBBIES_ASYNC action to the store with hobbies
    yield put({ type: GET_HOBBIES_ASYNC, hobbies });
  } catch (error) {
    // dispatches GET_REQUEST_FAILED action to the store with the error
    yield put({ type: GET_REQUEST_FAILED, error });
  }
}

// watcher saga: watches for GET_HOBBIES action dispatched to the store, starts worker saga
export function* watchGetHobbiesAsync() {
  yield takeEvery(GET_HOBBIES, getHobbiessAsync);
}
