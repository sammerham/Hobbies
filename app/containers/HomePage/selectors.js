import { createSelector } from 'reselect';

// hobbies selector
const hSelector = state => state.hobbies.hobbies;
// error selector
const eSelector = state => state.hobbies.error;

// creates memoized selector functions that only recalculate the output if the inputs change.
export const hobbySelector = createSelector(
  hSelector,
  hobbies => hobbies,
);
export const errorSelector = createSelector(
  eSelector,
  error => error,
);
