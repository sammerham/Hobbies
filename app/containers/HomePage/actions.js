/*
 *
 * NewHobbyPage actions
 *
 */

import { GET_HOBBIES } from './constants';
// dispatched when hobby added
export function getHobbies() {
  return {
    type: GET_HOBBIES,
  };
}
