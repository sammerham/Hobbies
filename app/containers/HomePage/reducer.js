import { GET_HOBBIES_ASYNC, GET_REQUEST_FAILED } from './constants';

const DEFAULT_STATE = {
  hobbies: [],
  error: false,
};

export default function hobbiesReducer(state = DEFAULT_STATE, action) {
  switch (action.type) {
    case GET_HOBBIES_ASYNC:
      return { hobbies: action.hobbies };
    case GET_REQUEST_FAILED:
      return { error: action.error };
    default:
      return state;
  }
}
