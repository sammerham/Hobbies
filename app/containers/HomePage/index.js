import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { hobbySelector, errorSelector } from './selectors';
import { getHobbies } from './actions';
import { StyledParag } from '../../components/Paragraph';
import ListItem from '../../components/ListItem';
/*
 * HomePage
 *
 * This is the first thing users see of our App, at the '/' route
 * Routes-- -> HomePage
 */

export default function HomePage() {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(getHobbies());
  }, []);
  const hobbies = useSelector(hobbySelector);
  const error = useSelector(errorSelector);

  if (error) {
    return <ListItem item="Something went wrong, please try again!" />;
  }
  if (hobbies.length === 0) {
    return <ListItem item="Hobby List is empty, enter your hobby" />;
  }
  if (hobbies.length) {
    return hobbies.map(hobby => (
      <StyledParag key={uuidv4()}>{hobby}</StyledParag>
    ));
  }
  return <FontAwesomeIcon icon={faSpinner} spin />;
}
