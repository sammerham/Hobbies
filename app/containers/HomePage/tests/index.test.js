import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { memoryHistory } from 'react-router-dom';
import { Provider } from 'react-redux';
import configureStore from '../../../configureStore';

import HomePage from '../index';
const store = configureStore({}, memoryHistory);

describe('<HomePage />', () => {
  it('should render and match the snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <IntlProvider locale="en">
        <Provider store={store}>
          <HomePage />
        </Provider>
      </IntlProvider>,
    );
    expect(firstChild).toMatchSnapshot();
  });
});
