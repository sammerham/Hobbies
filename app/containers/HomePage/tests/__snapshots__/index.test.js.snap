// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render and match the snapshot 1`] = `
<li
  class="Wrapper-euo0oy-0 jLvxtc"
>
  <div
    class="Item-sc-3y9mie-0 fKXSkT"
  >
    Hobby List is empty, enter your hobby
  </div>
</li>
`;
